File: src/spoolmanager.client/src/models/domain/Spool.ts
import { Filament } from './Filament';

export interface PrintJobSpool {
  id: number;
  printJobId: number;
  gramsUsed: number;
}

export interface Spool {
  id: number;
  filamentId: number;
  filament?: Filament;
  firstUsed: string | null;
  lastUsed: string | null;
  initialWeight: number;
  usedWeight: number;
  lotNumber: string;
  price: number;
  registered: string;
  spoolWeight: number;
  location: string;
  comment: string;
  costPerGram?: number;
  remainingGrams?: number;
  showName?: string;
  printJobSpools?: PrintJobSpool[];
}

File: src/SpoolManager.Server/Domain/Entities/Spool.cs
namespace SpoolManager.Server.Domain.Entities
{
    public class Spool
    {
        public int Id { get; set; }
        public string Comment { get; set; }
        public int FilamentId { get; set; }
        public DateTimeOffset? FirstUsed { get; set; }
        public decimal InitialWeight { get; set; }
        public decimal UsedWeight { get; set; }
        public DateTimeOffset? LastUsed { get; set; }
        public string Location { get; set; }
        public string LotNumber { get; set; }
        public decimal Price { get; set; }
        public DateTimeOffset Registered { get; set; }
        public decimal SpoolWeight { get; set; }
        public Filament Filament { get; set; }

        public decimal CostPerGram => Price / InitialWeight;

        public decimal RemainingGrams => InitialWeight - PrintJobSpools.Sum(pjf => pjf.GramsUsed);

        public string ShowName => $"{Filament.Brand.Name} - {Filament.Material} -  {Filament.Name}";
        public ICollection<PrintJobSpool> PrintJobSpools { get; set; } = new List<PrintJobSpool>();
    }
}
